packages/dds/tree/src/simple-tree/tree.ts

```diff
@@ -131,30 +198,156 @@ export class TreeViewConfiguration<TSchema extends ImplicitFieldSchema = Implici
+			}
+		} else if (schema.kind === NodeKind.Array) {
+			arrays.push(schema);
+		} else {
+			assert(schema.kind === NodeKind.Map, "invalid schema");
+			maps.push(schema);
+		}
+	}
+
+	if (arrays.length > 1) {
+		errors.push(
+			`More than one kind of array allowed within union (${formatTypes(arrays)}). This would require type disambiguation which is not supported by arrays during import or export.`,
+		);
+	}
+
+	if (maps.length > 1) {
+		errors.push(
+			`More than one kind of map allowed within union (${formatTypes(maps)}). This would require type disambiguation which is not supported by maps during import or export.`,
+		);
+	}
+
+	if (maps.length > 0 && arrays.length > 0) {
+		errors.push(
+			`Both a map and an array allowed within union (${formatTypes([...arrays, ...maps])}). Both can be implicitly constructed from iterables like arrays, which are ambiguous when the array is empty.`,
+		);
+	}
+
+	if (objects.length > 0 && maps.length > 0) {
+		errors.push(
+			`Both a object and a map allowed within union (${formatTypes([...objects, ...maps])}). Both can be constructed from objects and can be ambiguous.`,
+		);
+	}
+
+	// Check for objects which fully overlap:
+	for (const schema of objects) {
+		// Check that each object has at least one required field which is unique to it.
+		// This is a conservative uniqueness check that ensures that its possible to disambiguate objects,
+		// even if they have extra non-field members on them (that don't collide with any fields in the union), without even checking the types of fields.
+
+		const possiblyAmbiguous = new Set(objects);
+
+		// A schema can't be ambiguous with itself
+		possiblyAmbiguous.delete(schema);
+
+		for (const [key, field] of schema.fields) {
+			if (field.kind === FieldKind.Required) {
+				const withKey = allObjectKeys.get(key) ?? fail("missing schema");
+				for (const candidate of possiblyAmbiguous) {
+					if (!withKey.has(candidate)) {
+						possiblyAmbiguous.delete(candidate);
+					}
+				}
+			}
+		}
+
+		if (possiblyAmbiguous.size > 0) {
+			// TODO: make this check more permissive.
+			// Allow using the type of the field to disambiguate, at least for leaf types.
+			// Add "constant" fields which can be used to disambiguate even more cases without adding persisted data: maybe make them optional in constructor?
+			// Consider separating unambiguous implicit construction format from constructor arguments at type level, allowing constructor to superset the implicit construction options (ex: optional constant fields).
+			// The policy here however must remain at least as conservative as shallowCompatibilityTest in src/simple-tree/toMapTree.ts.
+
+			errors.push(
+				`The required fields of ${JSON.stringify(schema.identifier)} are insufficient to differentiate it from the following types: ${formatTypes(possiblyAmbiguous)}. For objects to be considered unambiguous, each must have required fields that do not all occur on any other object in the union.`,
```

Comment:
I wonder if the following might be a bit easier to parse:
"For objects to be considered unambiguous, each must have at least one required field that does not occur in any other object in the union."

Reply:
That used to be the policy, but the current policy is more relaxed than that. For example an objects with required fields `{a, b}`, `{b, c}` and `{c, a}` are permitted by this check and my phrasing.

 That said the comment the top of the loop above is out of date, which is likely the source of confusion. I'll update that.

Reply:
Note that the old policy accidently disallowed cases like a union of only one type, which was an object node with no fields.

Reply:
Comment fixed.